Run bench on demo

ffakenz opened the pull request:
<!-- Describe your change here -->

This PR enhances the [demo tutorial](https://hydra.family/head-protocol/docs/getting-started/) by enabling `hydra-cluster` benchmarks to run on an active Hydra cluster.

**usage**

```sh
nix run .#legacyPackages.x86_64-linux.hydra-cluster.components.benchmarks.bench-e2e -- \
          demo \
          --output-directory=$(pwd)/benchmarks \
          --scaling-factor=100 \
          --timeout=1000s \
          --testnet-magic 42 \
          --node-socket=${NETWORK_DIR}/node.socket \
          --hydra-client=localhost:4001 \
          --hydra-client=localhost:4002 \
          --hydra-client=localhost:4003
```

**prerequisites**
- A Cardano node must be running on specified `node-socket`.
- Hydra nodes must be operational on provided `hydra-client` hosts.
- There’s no need to pre-seed the keys, as the bench-demo script will automatically fund them using the faucet.
- Note that the reference scripts should already be published, and the Hydra nodes must be running with those scripts.


**Todo**

- [x] Fix the `FIXME` about `> 33`
- [x] Remove duplicate seeding
- [x] Make sure the entire CI process doesn't fail when the pumba causes the network to fail
- [x] Make it so that if it _fails_ the head is closed.
- [x] Quick little matrix to run a few different scenarios
- [x] Make the bench-e2e fail if it didn't submit all the txns ( ideally would also be able to see visually in the job list; but Github is missing a feature see also https://github.com/actions/runner/issues/2347 )
- [x] Get docker info via `docker inspect` instead of parsing yaml (!)
- [x] Make sure `results.csv` is written to the `outputDirectory` not the tmp directory
- [x] Upload the results as part of the artifacts
- [x] Write the summary out even when it failed

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
Define basic demo options

ffakenz committed:
Extract genDatasetConstantUTxO

To take faucet and client keys as arg,
so that is reusable also for the current use-case.

ffakenz committed:
Extract bench scenario

So that is also reusable for the current use-case.

ffakenz committed:
Extract findRunningCardanoNode'

To find a node runing on unknown network, like devnet.

So that is also reusable for the current use-case.

ffakenz committed:
Add network id as part of the options

Also removed starting node id as not needed,
because Alice will always lead the demo bench.

ffakenz committed:
Draft base bench demo script

ffakenz committed:
Fix seeding devnet on bench-demo given scripts were already published

ffakenz committed:
Fix tx building for  funding transaction

This solves the problem of:

SubmitTxValidationError (TxValidationErrorInCardanoMode (ShelleyTxValidationError ShelleyBasedEraBabbage (ApplyTxError (UtxowFailure (UtxoFailure (AlonzoInBabbageUtxoPredFailure (FeeTooSmallUTxO (Coin 169153) (Coin 0)))) :| []))))

The problem was we were using a build-raw transaction api instead of one which autobalances it.

ffakenz committed:
Add hydra signing keys as option args to bench-demo

ffakenz committed:
Extend generator of constant utxo property

To be valid for an arbitrary set of client keys,
instead of just a single one.

noonio committed:
Trying to work out why large numbers of clients fail

- Also note strange keys generated when allocating only small amounts of
  lovelace

ffakenz committed:
Limit the nbr of client keys on arbitrary for dataset

ffakenz committed:
Return funds to faucet when it finishes

ch1bo commented on hydra-cluster/bench/Bench/Options.hs:
```diff
@@ -156,6 +168,38 @@ startingNodeIdParser =
+          ( "Hydra signing key of a party in the Head. Can be \
+            \provided multiple times, once for each participant (current maximum limit is "
+              <> show maximumNumberOfParties
+              <> " )."
```
Are these really hydra signing keys?

Why is there a maximum of keys we can provide?

ffakenz replied:
this assumes you have a hydra-cluster on  at most `maximumNumberOfParties ` running on the background and the script will act on behalf of the user to commit, etc..

ch1bo commented on hydra-cluster/bench/Bench/Options.hs:
```diff
@@ -156,6 +168,38 @@ startingNodeIdParser =
+  info
+    demoOptionsParser
+    ( progDesc
+        "Run scenarios from local running demo."
```
This requires three hydra nodes listening on ports 4001, 4002 and 4003.

Should: document this assumption somewhere.

ch1bo commented on hydra-cluster/bench/Bench/EndToEnd.hs:
```diff
@@ -95,57 +96,116 @@ bench startingNodeId timeoutSeconds workDir dataset@Dataset{clientDatasets, titl
+              putTextLn "Seeding network"
+              fundClients networkId nodeSocket fundingTransaction
+              forM_ clientSks (fuelWith100Ada (contramap FromFaucet tracer) node)
+              putStrLn $ "Connecting to hydra cluster in " <> workDir
```
The cluster is not running in a work dir. Instead, we are connecting to the three (expected to run) hydra nodes on ports 4001, 4002 and 4003 here.

Should: put that on the stdout instead

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Faucet.hs:
```diff
@@ -106,19 +106,30 @@ returnFundsToFaucet ::
+returnFundsToFaucet' ::
+  Tracer IO FaucetLog ->
+  RunningNode ->
+  VerificationKey PaymentKey ->
```
This is never different to the `keysFor Faucet`. 

Should: remove this argument

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Faucet.hs:
```diff
@@ -106,19 +106,30 @@ returnFundsToFaucet ::
+  Tracer IO FaucetLog ->
+  RunningNode ->
+  VerificationKey PaymentKey ->
+  (VerificationKey PaymentKey, SigningKey PaymentKey) ->
```
Could: only pass `SigningKey PaymentKey`

you can always get the verification key from the signing key, while we always require the signing key. This should make call sites simpler

ch1bo commented on hydra-cluster/src/Hydra/Generator.hs:
```diff
@@ -130,3 +125,59 @@ genDatasetConstantUTxO faucetSk nClients nTxs = do
+
+genDatasetConstantUTxODemo ::
+  -- | The faucet keys
+  (VerificationKey PaymentKey, SigningKey PaymentKey) ->
```
Why do we need to pass in the keys where we know that funds must be available at the `Faucet`?

Should: not need this.

ch1bo commented on hydra-cluster/src/Hydra/Generator.hs:
```diff
@@ -130,3 +125,59 @@ genDatasetConstantUTxO faucetSk nClients nTxs = do
+  NetworkId ->
+  SocketPath ->
+  IO Dataset
+genDatasetConstantUTxODemo (faucetVk, faucetSk) allClientKeys nTxs networkId' nodeSocket = do
```
Have you tried this function against the freshly seeded Head? I think it would just work too and might be deterministic .. so maybe no need for `mkGenesisTx`. Alternatively, doing the querying outside and passing in a `UTxO` owned by the faucet to be spent should allow us to re-use more code.

Should: not duplicate the funding transaction logic

ch1bo commented on hydra-cluster/src/Hydra/Generator.hs:
```diff
@@ -87,34 +94,22 @@ generateConstantUTxODataset ::
-  -- | Number of clients
-  Int ->
+  -- | Clients
+  [ClientKeys] ->
```
Should: not use `ClientKeys` here. This type is strictly bigger than what we need. A list of cardano signing keys should be enough here. i.e. `[SigningKey PaymentKey]`

ffakenz replied:
The problem is that the function has to Generate a Dataset and for that it needs `[ClientKeys]`

So from my point of view is unavoidable. Is there anything I'm missing here?

ch1bo replied:
The comment was about that `ClientKeys` contains two signing keys. But we only need one of them (iirc it was called `externalSigningKey`) to fund the client.

ch1bo commented on hydra-cluster/test/Test/GeneratorSpec.hs:
```diff
@@ -38,21 +34,31 @@ spec = parallel $ do
 prop_keepsUTxOConstant =
-  forAll arbitrary $ \(Positive n) -> do
+  -- FIXME: Breaks with values > 33. Why?
+  forAll (inputs maximumNumberOfParties) $ \(Positive n, clientKeys) -> do
```
Must: address this. Nothing should stop us from having more than `maximumNumberOfParties` clients in a benchmark!

The current benchmark just was always the same number of participants as clients, but in "demo mode", we do not care how big the cluster of hydra nodes actually is.

ch1bo commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -132,26 +132,30 @@ getCardanoNodeVersion =
+
+  CardanoNodeArgs{nodeSocket} = defaultCardanoNodeArgs
+
+findRunningCardanoNode' :: Tracer IO NodeLog -> NetworkId -> SocketPath -> IO (Maybe RunningNode)
```
Should: add haddocks

ch1bo commented on hydra-cluster/src/CardanoClient.hs:
```diff
@@ -111,25 +111,22 @@ waitForUTxO node utxo =
       error $ "Unexpected TxOut " <> show txOut
 
-mkGenesisTx ::
+mkInitialTx ::
```
This is NOT a good name. We have a protocol transaction that is exactly called this.

Also, the error call still calls this "genesis transaction".

Suggestion: `buildRaw` is only used in this function. Let's move it into this function and call this `buildRawTransaction` or so. Adding a haddock could also help to explain what the difference between this, `buildTransaction` and `mkGenesisTx` is (if we even keep the latter).

ch1bo commented on hydra-cluster/src/CardanoClient.hs:
```diff
@@ -111,25 +111,22 @@ waitForUTxO node utxo =
   -- | Recipients and amounts to pay in this transaction.
   [(VerificationKey PaymentKey, Coin)] ->
+  TxIn ->
+  -- | Initial input from which to spend
```
Haddock referencing wrong parameter. Also, it would be more logical to put this after `NetworkId` and before other arguments which refer the "things to spend", while the "things to produce" come after

ch1bo commented on hydra-cluster/bench/Bench/EndToEnd.hs:
```diff
@@ -238,20 +298,22 @@ movingAverage confirmations =
+fundClients networkId nodeSocket fundingTransaction = do
+  putTextLn "Fund scenario from faucet"
+  submitTransaction networkId nodeSocket fundingTransaction
+  void $ awaitTransaction networkId nodeSocket fundingTransaction
```
Having this in a separate function is pointless. The `putTextLn` will rarely fit and it's basically an alias for `submitTransaction >=> void . awaitTransaction`

Should: keep `fuelWith100Ada` local to `seedNetwork` and instead just submit and await where you called `fundClients`

ch1bo commented on hydra-cluster/bench/Bench/EndToEnd.hs:
```diff
@@ -95,57 +96,116 @@ bench startingNodeId timeoutSeconds workDir dataset@Dataset{clientDatasets, titl
+  send leader $ input "Init" []
+  headId <-
+    waitForAllMatch (fromIntegral $ 10 * clusterSize) clients $
+      headIsInitializingWith parties
```
This is the only location where you actually use the `--hydra-sk` options. Instead of going at lengths to provide them, we could just not care too much about what head we see initializing. For example we could just check it's a head with the expected number of participants (`length (leader:followers)`)

Should: not need `--hydra-sk` just for this reason.

ch1bo reviewed (changes requested):
I like that you have put a pull request description explaining how to use this.

However, it would be better (maybe even required) that you leave instructions somewhere on the repository that this mode exists and how to use it! Potential places:
- getting started documentation page (maybe not as this is a first touch point for people
- hydra-cluster/README.md which also explains the other running modes of the benchmarks: https://github.com/input-output-hk/hydra/blob/1b0111f0bd23349e1381a7952b38d22ff1386b24/hydra-cluster/README.md#L114 (probably best to explain it here)

Furthermore, I'm not too happy about where you "cut" and duplicated the existing dataset generation logic. I think we can re-use much more code if we not distinguish too and just fetch the faucet utxo before generating the UTxO. In the clean devnet case the dataset should still be deterministic (that's why we did do it with genesis utxo originally).

I see quite some unnecessary complexity also coming from the fact that we have `Dataset` unchanged, while in demo mode we don't actually need the "fuel keys" in the `ClientKeys` (actually we also don't need the hydra keys, see other comments). Our assumption is (and can be) that the nodes we connect to have access to the right keys and have some fuel to spend.

ffakenz committed:
Avoid passing network id to options and faucet keys on generators

As we forcing the bench to run on devnet for the moment.

ffakenz committed:
Make bench-demo sim to run on top of previous run

For that we made the following changes:
- avoid requesting for node's history upon connection
- datasets transactions use sender same as recipient
- finally we avoid waiting for nodes to be connected given that is assumed in this case

ch1bo committed:
Start a github action for network testing

ch1bo committed:
Spin up things individually

ch1bo committed:
Acquire and upload logs

ch1bo committed:
Make the hnode fallback in demo/seed-devnet.sh non-interactive tty

ch1bo committed:
Add invocation to produce traffic

noonio committed:
check socket permissions

noonio committed:
ls not la

noonio committed:
tmate

noonio committed:
tmate; socket file permissions

noonio committed:
split setup and run benchmarks

noonio committed:
Run pumba and the benchmarks

- Build and run seperate nix steps

noonio committed:
Line buffering; 100% loss on pumba

noonio committed:
tmate

ffakenz committed:
Replace hydra-keys options by hydra-client hosts instead

Given we connect to provided hydra-clients we don't need to check
we observe a head is inicializing with the same parties as its assumed.
Instead we take hydra-client hosts so that the solution scales in other settings.

ffakenz committed:
Rebase and make use new hydra-client options during CI

noonio committed:
Use docker version of pumba

It seems to have the right priviledges to do the netem work

noonio committed:
Does it work with 20% loss?

noonio committed:
See if it succeeds with 0% loss

noonio committed:
What about 5%

noonio committed:
Optional tmate'ing

noonio committed:
Disable condition as it doesn't work

noonio committed:
Try nix again; leave very long running time

noonio committed:
Try a sleep

noonio committed:
Horrible hack

noonio committed:
Note about checking for pumba working

noonio committed:
Target node 1 for loss and limit target ips for node-2, and node-3

noonio committed:
Try /24

noonio committed:
Multiple targets is right

noonio committed:
Let's try 2%

noonio committed:
Add now 5%

noonio committed:
3%

noonio committed:
4%

ffakenz committed:
Try to wait for alice peer disconnected from bob and carol nodes

noonio committed:
Executable; compatible with NixOS

ffakenz committed:
Add persitency-dir options to hydra-node

Due to the watch_logs script isn't able to find it

noonio committed:
Use a special image built for netem; no more tcimage

ffakenz committed:
Add base network testing readme docs

ffakenz committed:
Use local functions to seed the network

Instead of exposing private ones defined in where clauses.

ffakenz committed:
Do not seed neetwork before runnin bench-demo

Because the script will be seeding the network.
The reason for this its because the scripts does not rely on
a genesis funding transaction.

ffakenz committed:
Remove FIXME as its not reproduceable

It just simply does not break ¯\_(ツ)_/¯

ffakenz committed:
Minor refactor on generateOneTransfer

So that is reusable for this use-case without braking others.

noonio committed:
Use random/self transfer generator functions

ffakenz committed:
Update demo-bech to take datasets from args

ffakenz committed:
Gen demo datasets

ffakenz committed:
Refactor HydraClient to keep peer information

This is useful to triger node actions not depending on localhost,
like in demo-bench where commit has to be done on running peers.

ffakenz committed:
dbg

ffakenz committed:
Try to use the pre-seeded client funds instead of faucet's

ffakenz committed:
Generate each client dataset from a different client funding tx

Instead of usign a single one from the faucet current utxo.

ffakenz committed:
Try not to specify any output

ffakenz committed:
Try using mkTxOutAutoBalance instead

ffakenz committed:
Fix client utxo

ffakenz committed:
Try not to specify any output ~ v2

ffakenz committed:
Go back to unbalanced outputs ~ v2

ffakenz committed:
Make funding tx optional for datasets

This is used to seed the network in other scenarios.
This one assumes the cluster is pre-seeded.

ffakenz committed:
Try not to specify any output ~ v3

ffakenz committed:
Try using mkTxOutAutoBalance instead ~ v2

noonio committed:
WIP; Revert faucet; use buildRawTransaction

noonio committed:
Create workdir if it doesn't exist

noonio committed:
WIP on generateDemoUTxO function; just missing witnesses.

noonio committed:
Small refactors around types

noonio committed:
WIP back to generating the dataset on the fly

ffakenz committed:
Do not pre-seed devnet

Instead, let the scenario seed the network and
provide funds to running clients.
At the end, funds are returned back to the faucet.

noonio committed:
Use other script for exporting pparams

noonio committed:
Short contestation period so that it closes fast

ffakenz committed:
add failure handling on tx processing while head is open

this allows us to close the head when we get a
time exceeded while wait for all confirmations when processing a tx.

noonio committed:
Compute pct of transactions confirmed

noonio committed:
4% should work

ffakenz committed:
Enhance workflow

ffakenz committed:
Some nice refactoring

ffakenz committed:
Make watch for logs optional

noonio committed:
Reframe wait for peer disconnected variable

noonio committed:
Experiment with a matrix

noonio committed:
Name in artifact

noonio committed:
Expect-failure matrix configuration

noonio committed:
Add failure mode we expect; refine matrix

noonio committed:
Fix order of detecting error conditions

noonio committed:
Add scaling factor to matrix config

noonio commented:
While Github doesn't show the specific matrix configuration fail, it _does_ show it as an "annotation"; so that's extremely useful for verifiying that the scenarios we expect fail in the right way.

![image](https://github.com/user-attachments/assets/d80388a8-aa1e-4b3f-8651-455cbd4eaada)

ffakenz committed:
Replace yq by docker inspect

ffakenz committed:
Upload the results as part of the artifacts

ffakenz committed:
Make sure results.csv is written to the outputDirectory not the tmp directory

ffakenz committed:
Write the summary out even when it failed

noonio commented:
This is looking really great!

noonio committed:
Make peers part of the matrix

noonio committed:
Allow 5% to fail

noonio committed:
Remove explicit failure tracking in the includes/excludes

noonio committed:
Add extra scaling factor

ch1bo commented on .github/workflows/network-test.yaml:
```diff
@@ -0,0 +1,158 @@
+          - alice
+          - bob
+          - carol
+        default: alice
```
The amount of bash scripting required (`peers_info_json.sh`) to determine the actual docker container IP addresses from this is baffling.

Does it actually make a difference which node we target with our fault injection?

Why are we not just using pumbas `--random` on docker container names from `docker-compose.yaml` we use?

noonio replied:
I think this will go away on a subsequent refactor of more advanced selection of the number of peers and which peers to target with the loss.

We can't use `--random` for two reasons: one, it blocks us a bit from loss on multiple peers; and it means we can't be clear about which traffic to which IP address to block.

ch1bo replied:
Then what about providing a list of container names? those would be directly evident from the used `docker-compose.yaml`, while the names are only indirectly connected and require the `peers_info_json.sh` script.

Alternative: rename the docker containers to `hydra-node-{alice, bob, carol}` and just use this parameter directly in the `run-pumba.sh`?

My point is: can we avoid needing `peers_info_json.sh`?

noonio replied:
I definitely think peers_info_json.sh will go away in the subsequent PR that will make the peer configuration more generic/testable

I personally don't think it's worth refactoring away just at this moment; I think we can do it in a subsequent PR.

I agree that that then means that technically we could delete it (and do a bit more refactoring to support that); but my opinion is that it doesn't hurt to have it at the moment; it's more work to delete, and it indicates a nice idea that we want to replicate, even if not done in exactly the same way.

I.e. I'm certain we'll refactor it away subsequently.

ffakenz replied:
This suggestion has been pushed into a red-bin item on our board:
https://github.com/orgs/cardano-scaling/projects/6?pane=issue&itemId=77063030

ch1bo commented on .github/workflows/network-test.yaml:
```diff
@@ -0,0 +1,156 @@
+        type: boolean
+        description: 'Check other peers see the target peer is disconnected before running benchmark.'
+        required: false
+        default: false
```
Why do we want to have peers be disconnected?

This also requires a fair bit of hard-to-maintain bash scripts that parse node log outputs. I think we should avoid that if possible?

noonio replied:
It's true that we're not using it right now; we could just delete this and keep the scripts around in some scratch space for later usage, potentially.

ffakenz replied:
This suggestion has been pushed into a red-bin item on our board:
https://github.com/orgs/cardano-scaling/projects/6?pane=issue&itemId=77063030

noonio committed:
Matrix information in the name

ch1bo commented on hydra-cluster/README.md:
```diff
@@ -140,3 +140,14 @@ The benchmark can be run in two modes corresponding to two different commands:
+`results.csv` file in a work directory. Same as for benchmarks results, you can use the `bench/plot.sh` script to plot the transaction confirmation times.
+
+To run the benchmark in this mode, the command is:
+* `demo`: Runs one or more preexisting _datasets_ in sequence and collect their results in a single markdown formatted file. The purpose of this setup is to facilitate a variaty of network-resiliance scenarios, such as packet loss or node failures. This is useful to prove the robustness and performance of the hydra-node's network over time and produce a human-readable summary.
```
Must: document this consistently.

It seems like the current code does not use "preexisting datasets" anymore.

ffakenz replied:
good catch!

ch1bo commented on hydra-cluster/bench/Bench/EndToEnd.hs:
```diff
@@ -97,55 +100,135 @@ bench startingNodeId timeoutSeconds workDir dataset@Dataset{clientDatasets, titl
+              forM_ clientSks $ \ClientKeys{signingKey} -> do
+                let vk = getVerificationKey signingKey
+                putTextLn $ "Seed client " <> show vk
+                seedFromFaucet node vk 100_000_000 (contramap FromFaucet tracer)
```
Should: DRY seeding

As now both, `bench` and `benchDemo`, do use the funding transaction, we should move the seeding of keys to the `scenario` or re-use the `seedNetwork` function.

ffakenz replied:
can't re-use the `seedNetwork` function because it calls `publishHydraScriptsAs`.

ch1bo replied:
Then move `publishHydraScriptsAs` out of it and only do that in the non-demo `bench`?

ch1bo commented on hydra-cluster/bench/Bench/EndToEnd.hs:
```diff
@@ -275,9 +358,12 @@ processTransactions clients Dataset{clientDatasets} = do
+        `concurrently_` progressReport (hydraNodeId client) clientId numberOfTxs submissionQ
+      )
+      `catch` \(_ :: SomeException) ->
+        putStrLn "Time exceeded while wait for all confirmations"
```
Is this the only exception possible here? Likely not, so maybe rephrase this to not put someone debugging on the wrong path.

ch1bo commented on hydra-cluster/bench/Bench/Options.hs:
```diff
@@ -156,6 +149,39 @@ startingNodeIdParser =
+        "Run bench scenario over local demo. \
+        \ This requires having in the background: \
+        \   * cardano node running on specified node-socket. \
+        \   * three hydra nodes listening on ports 4001, 4002 and 4003."
```
Must: fix this description

It seems like the command can be parameterized by the number of nodes and which ports are used via `--hydra-client` options.

ffakenz replied:
good catch!

ch1bo commented on hydra-cluster/bench/Main.hs:
```diff
@@ -95,15 +137,15 @@ benchmarkFailedWith benchDir = \case
 
   writeReport outputDir = do
-    let reportPath = outputDir </> "end-to-end-benchmarks.md"
+    let reportPath = outputDir </> (maybe "" (<> "-") lbl <> "end-to-end-benchmarks.md")
```
Does this affect the slug of https://hydra.family/head-protocol/benchmarks/end-to-end-benchmarks for the benchmarks done in CI?

ch1bo commented on hydra-cluster/bench/Bench/EndToEnd.hs:
```diff
@@ -97,55 +100,135 @@ bench startingNodeId timeoutSeconds workDir dataset@Dataset{clientDatasets, titl
+    mapM_
+      ( \sender -> do
+          returnAmount <- returnFundsToFaucet' faucetTracer node sender
+          traceWith faucetTracer $ ReturnedFunds{actor = show sender, returnAmount}
```
Note that this logs a signing key. Although this is a benchmark / test, this would be quite a bad thing to do in production!

Also: why not trace inside `returnFundsToFaucet'` as it already takes the `faucetTracer`?

ffakenz replied:
Good catch!

Loging verification keys inside `returnFundsToFaucet'` instead.

ch1bo commented on hydra-cluster/src/Hydra/Generator.hs:
```diff
@@ -80,53 +82,108 @@ defaultProtocolParameters = def
+      Left e -> throwIO $ FaucetFailedToBuildTx{reason = e}
+      Right body -> do
+        let signedTx = sign faucetSk body
+        pure signedTx
```
Could: DRY this with building the funding transaction in `generateConstantUTxODataset`.

We have two variants of dataset returning functions now:
* `generateConstantUTxODataset` uses hard-coded `availableInitialFunds`, `Faucet` keys and generates clients keys
* `generateDemoUTxODataset` that queries available funds takes keys as argument

This could be combined differently such that we could define one in terms of the other. Also, do we still need the hard-coded variant if we would query the devnet there too?

ffakenz replied:
good idea.

lets address this in a subsequent refactor:

https://github.com/orgs/cardano-scaling/projects/6?pane=issue&itemId=77063030

ch1bo commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -401,8 +402,15 @@ withHydraNode' tracer chainConfig workDir hydraNodeId hydraSKey hydraVKeys allNo
+  withConnectionToNodeHost tracer hydraNodeId Host{hostname, port}
+ where
+  hostname = "127.0.0.1"
+  port = fromInteger $ 4_000 + toInteger hydraNodeId
```
Could: just get rid of the `Int` variant and push the `4_000 + toInteger hydraNodeId` further upstream? or into a function `testApiHostFromNodeId :: Int -> Host`

ffakenz replied:
good idea.

lets address this in a subsequent refactor:

https://github.com/orgs/cardano-scaling/projects/6?pane=issue&itemId=77063030

ch1bo commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -172,42 +173,42 @@ waitForAll tracer delay nodes expected = do
 -- this parses the response as a 'UTxO' type as we often need to pick it apart.
 getSnapshotUTxO :: HydraClient -> IO UTxO
-getSnapshotUTxO HydraClient{hydraNodeId} =
+getSnapshotUTxO HydraClient{peer = Host{hostname, port}} =
```
Must: must not call this `peer`, but `apiHost` or so (to avoid confusion)

It's _not_ a "peer" what we configure the hydra-node with (`--peer`), but it points to the API of a hydra node.

ch1bo reviewed (changes requested):
I like that we use github action matrix to run multiple scenarios now.

I do not like the amount of shell scripting in coordination of `pumba`. Some of which could be avoided IMO (see individual comments).

On the `bench-e2e` binary the re-use of code is much better now and only small inconsistencies _must_ be addressed, while I left suggestions that _should_ or _could_ be done (now or later).

ffakenz committed:
Rename hydra clients peers to api hosts

ffakenz committed:
Update README

ffakenz committed:
Enhance user error during processTransactions

ffakenz committed:
Update bench demo option description

ffakenz committed:
Make the scenario to seed the network

ffakenz committed:
Rever changes over the benchmarks slug used by the website

ffakenz committed:
Do not trace signing keys during benchmark

ffakenz committed:
Remove watch logs step and simplify workflow args

ch1bo commented on .github/workflows/network-test.yaml:
```diff
@@ -0,0 +1,133 @@
+
+        peers_info_json=$(.github/workflows/network/peers_info_json.sh)
+
+        .github/workflows/network/run_pumba.sh $target_peer $percent $peers_info_json
```
Should: get rid of the indirection through `peers_info_json.sh` as we know the container names from line 60. The target peer is even constant now?

ch1bo commented on hydra-cluster/bench/Bench/EndToEnd.hs:
```diff
@@ -86,66 +89,143 @@ bench startingNodeId timeoutSeconds workDir dataset@Dataset{clientDatasets, titl
+    waitForAllMatch (fromIntegral $ 10 * clusterSize) clients $
+      if null parties
+        then aHeadIsInitializingWith (length clients)
+        else headIsInitializingWith parties
```
Could: just not do match for the exact number of parties (or length of clients).

Then we also don't need `aHeadIsInitializingWith`

ch1bo commented on hydra-cluster/hydra-cluster.cabal:
```diff
@@ -221,6 +221,7 @@ benchmark bench-e2e
     , regex-tdfa
     , scientific
     , statistics
+    , text
```
Should: not need this as our prelude re-exports quite a bit from `text`. What was this added for?

ffakenz replied:
To `pack $ printf` on Summary

ch1bo commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -401,8 +402,15 @@ withHydraNode' tracer chainConfig workDir hydraNodeId hydraSKey hydraVKeys allNo
 
-withConnectionToNode :: forall a. Tracer IO HydraNodeLog -> Int -> (HydraClient -> IO a) -> IO a
-withConnectionToNode tracer hydraNodeId action = do
+withConnectionToNode :: forall a. Tracer IO HydraNodeLog -> Int -> Bool -> (HydraClient -> IO a) -> IO a
```
Should: avoid `Bool` arguments (boolean blindness)

But not going to pick a fight / require a refactor right away.

ffakenz replied:
agree, it was a hack.

now we passing maybe query params, which i think is more accurate.

wdyt?

ffakenz committed:
Remove unnecessary scripts

ffakenz committed:
Enhance failure message for HUnitFailure errors

ffakenz committed:
Remove hack to check every node sees head is initialized for demo

In this case the parties are unknown but we still need
to fetch the headId from a head id observation.

ffakenz committed:
Refactor hacky boolean blindness to manage query params on withConnectionToNode

ffakenz committed:
Fix not seeding during scenario but before it gets executed

The reason is  depends on  before starting the hydra-cluster.
That consumes the Faucet UTxO making the fundingTransaction from dataset invalid, thus
making the scenario to fail.

noonio committed:
Don't call --random on pumba; note about time
